Library/Homebrew/unpack_strategy.rb

```diff
@@ -0,0 +1,340 @@
+      SubversionUnpackStrategy,
+      CvsUnpackStrategy,
+      FossilUnpackStrategy,
+      BazaarUnpackStrategy,
+      P7ZipUnpackStrategy,
+      RarUnpackStrategy,
+      LhaUnpackStrategy,
+    ].freeze
+  end
+  private_class_method :strategies
+
+  def self.detect(path)
+    magic_number = if path.directory?
+      ""
+    else
+      File.binread(path, MAX_MAGIC_NUMBER_LENGTH)
+    end
+
+    strategy = strategies.detect do |s|
+      s.can_extract?(path: path, magic_number: magic_number)
+    end
+
+    # This is so that bad files produce good error messages.
+    strategy ||= case path.extname
+    when ".tar.gz", ".tgz", ".tar.bz2", ".tbz", ".tar.xz"
+      TarUnpackStrategy
+    when ".zip"
+      ZipUnpackStrategy
+    else
+      UncompressedUnpackStrategy
+    end
+
+    strategy.new(path)
+  end
+
+  attr_reader :path
+
+  def initialize(path)
+    @path = Pathname(path).expand_path
+  end
+
+  def extract(to: nil, basename: nil)
+    unpack_dir = Pathname(to || Dir.pwd).expand_path
+    unpack_dir.mkpath
+    extract_to_dir(unpack_dir, basename: basename)
+  end
+
+  def pipe_to_tar(tool, unpack_dir)
+    Utils.popen_read(tool, "-dc", path) do |rd|
+      Utils.popen_write("tar", "-x", "-f", "-", "-C", unpack_dir) do |wr|
+        buf = ""
+        wr.write(buf) while rd.read(16384, buf)
+      end
+    end
+  end
+  private :pipe_to_tar
+
+  def extract_to_dir(*)
+    raise NotImplementedError
+  end
+  private :extract_to_dir
+end
+
+class DirectoryUnpackStrategy < UnpackStrategy
+  def self.can_extract?(path:, **)
+    path.directory?
+  end
+
+  def extract_to_dir(unpack_dir, **)
+    FileUtils.cp_r path.children, unpack_dir, preserve: true
+  end
+  private :extract_to_dir
+end
+
+class UncompressedUnpackStrategy < UnpackStrategy
+  def self.can_extract?(**)
+    false
+  end
+
+  def extract_to_dir(unpack_dir, basename:)
+    FileUtils.cp path, unpack_dir/basename, preserve: true
+  end
+  private :extract_to_dir
+end
+
+class JarUnpackStrategy < UncompressedUnpackStrategy
+  def self.can_extract?(path:, magic_number:)
+    return false unless ZipUnpackStrategy.can_extract?(path: path, magic_number: magic_number)
+
+    # Check further if the ZIP is a JAR/WAR.
+    Open3.popen3("unzip", "-l", path) do |stdin, stdout, stderr, wait_thr|
+      stdin.close_write
+      stderr.close_read
+
+      begin
+        return stdout.each_line.any? { |l| l.match?(%r{\s+META-INF/MANIFEST.MF$}) }
+      ensure
+        stdout.close_read
+        wait_thr.kill
+      end
+    end
+  end
+end
+
+class P7ZipUnpackStrategy < UnpackStrategy
+  def self.can_extract?(magic_number:, **)
+    magic_number.match?(/\A7z\xBC\xAF\x27\x1C/n)
+  end
+
+  def extract_to_dir(unpack_dir, **)
+    safe_system "7zr", "x", "-y", "-bd", "-bso0", path, "-o#{unpack_dir}"
+  end
+  private :extract_to_dir
+end
+
+class ZipUnpackStrategy < UnpackStrategy
+  def self.can_extract?(magic_number:, **)
+    magic_number.match?(/\APK(\003\004|\005\006)/n)
```

Comment:
Feels like there's maybe a nicer interface for this given pretty much all the `self.can_extract` methods are the same?

Reply:
I can't think of any, since some need more than just the magic number.

Reply:
How about the default implementation in `UnpackStrategy` being `magic_number.match?(magic_regex)` or something and each subclass can define `magic_regex`?

Reply:
Tried it, but it only get's worse.

For example, `XzUnpackStrategy < TarUnpackStrategy < FileUnpackStrategy`:

This needs to override the custom `can_extract?` in `TarUnpackStrategy`, so it will be exactly the same as in `FileUnpackStrategy` again, plus also having to add a `magic_regex` constant/method.

Reply:
Cool, no worries, thanks for checking.